src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs

```diff
@@ -9,295 +9,445 @@
                 {
-                    if (usedLabels.Contains(label.Label))
-                    {
-                        if (sectionBuilder == null)
-                        {
-                            sectionBuilder = ArrayBuilder<BoundStatement>.GetInstance();
-                        }
-
-                        sectionBuilder.Add(_factory.Label(label.Label));
-                    }
+                    sectionBuilder.Add(_factory.Label(label.Label));
                 }
 
-                if (sectionBuilder != null)
-                {
-                    sectionBuilder.AddRange(VisitList(section.Statements));
-                    sectionBuilder.Add(_factory.Goto(node.BreakLabel));
-                    result.Add(_factory.Block(section.Locals, sectionBuilder.ToImmutableAndFree()));
-                }
+                // Add the translated body of the switch section
+                sectionBuilder.AddRange(VisitList(section.Statements));
+                sectionBuilder.Add(_factory.Goto(node.BreakLabel));
+                result.Add(_factory.Block(section.Locals, sectionBuilder.ToImmutableAndFree()));
+                // at this point the end of result is unreachable.
             }
 
             result.Add(_factory.Label(node.BreakLabel));
-            return _factory.Block(usedTemps.ToImmutableArray().Concat(node.InnerLocals), node.InnerLocalFunctions, result.ToImmutableAndFree());
+            var translatedSwitch = _factory.Block(pslr.DeclaredTemps.ToImmutableArray().Concat(node.InnerLocals), node.InnerLocalFunctions, result.ToImmutableAndFree());
+            return translatedSwitch;
         }
 
-        private void LowerDecisionTree(DecisionTree decisionTree, HashSet<LabelSymbol> usedLabels, HashSet<LocalSymbol> usedTemps, ArrayBuilder<BoundStatement> result)
+        private class PatternSwitchLocalRewriter
         {
-            if (decisionTree == null) return;
-            switch (decisionTree.Kind)
+            public readonly LocalRewriter LocalRewriter;
+            public readonly HashSet<LocalSymbol> DeclaredTempSet = new HashSet<LocalSymbol>();
+            public readonly ArrayBuilder<LocalSymbol> DeclaredTemps = ArrayBuilder<LocalSymbol>.GetInstance();
+            public readonly Dictionary<BoundPatternSwitchSection, ArrayBuilder<BoundStatement>> SwitchSections = new Dictionary<BoundPatternSwitchSection, ArrayBuilder<BoundStatement>>();
+
+            private ArrayBuilder<BoundStatement> _loweredDecisionTree = ArrayBuilder<BoundStatement>.GetInstance();
+            private readonly SyntheticBoundNodeFactory _factory;
+
+            public PatternSwitchLocalRewriter(LocalRewriter localRewriter, BoundPatternSwitchStatement node)
+            {
+                this.LocalRewriter = localRewriter;
+                this._factory = localRewriter._factory;
+                foreach (var section in node.SwitchSections)
+                {
+                    SwitchSections.Add(section, new ArrayBuilder<BoundStatement>());
```

Comment:
Because we have enough of them in flight that we'd probably overflow and underflow the pool. But it couldn't hurt; I'll do that.

---
In reply to: [70325086](https://github.com/dotnet/roslyn/pull/12416#discussion_r70325086) [](ancestors = 70325086)